code/modules/mob/living/carbon/carbon.dm

```diff
@@ -96,6 +96,59 @@
 
 	return
 
+//EMP vulnerability for non-synth carbons. could be useful for diona, vox, or others
+//the species' emp_sensitivity var needs to be greater than 0 for this to proc, and it defaults to 0 - shouldn't stack with prosthetics/fbps in most cases
+//higher sensitivity values incur additional effects, starting with confusion/blinding/knockdown and ending with increasing amounts of damage
+//the degree of damage and duration of effects can be tweaked up or down based on the species emp_mod and emp_stun_mod vars (default 1) on top of tuning the random ranges
+/mob/living/carbon/emp_act(severity)
+	if(species.emp_sensitivity > 0) //receive basic effects regardless of sensitivity, if we're even the tiniest bit sensitive
+		to_chat(src, "<span class='danger'><B>*BZZZT*</B></span>")
+		switch(severity)
+			if(1) //highest intensity, absolutely sucks
```

Comment:
magic numbers should be #defines, for clarity
And possibly use bitflags since these seem to control different factors.

Reply:
I'd contemplated bitflags or something like how species flags are handled (such as NO_POISON/NO_SCAN), but they're not something I've really messed with before so I was hesitant to start messing with things I didn't understand.

As for defines; I assume you're referring to the damage values specifically, and possibly effect durations?

Reply:
The numbers you're switching on. They don't mean anything to anyone else if nothing is written down (hence the phrase "magic numbers", their meaning is indecipherable i.e., magical, on account of the possibilities that the switching variable is computed, rather than stored somewhere, or that the selfsame variable is used elsewhere to different effect), so they're particularly likely to be misused by others who don't understand what precisely they mean. If you were to replace those with #defines, then you can elaborate on the meaning of the numbers in the name of the define, such that someone else can glance at the code and realize, "Ok, they're switching between these different modes of operation for this specific thing". Actually properly reading what it's switching on, this would work a lot like the event severity defines (`EVENT_LEVEL_MUNDANE`, `EVENT_LEVEL_MODERATE `, and `EVENT_LEVEL_MAJOR`), which pretty clearly tell you what the value is in reference to the involved systems. It's an event, and its severity is of some relative measurement.

Bitflags are pretty straightforward, so long as you remember to use logical AND/OR (&/|) instead of boolean (or short-circuited) AND/OR (&&/||). Basically, in a 32-bit integer variable (byond is notably 24-bit), you have 32 individual bits that can be either TRUE or FALSE. If you assign a particular meaning to each individual bit, then you can condense a set of up to 32 toggleable things into a single int, instead of requiring 32 separate boolean variables to represent each individual toggle, or trying to use complex switching logic to encapsulate various combinations of the different toggles.

A simple example of why they're useful is in this file, where you have a few if-else if blocks, where you've got something like `if(species.emp_sensistivity == 3 || species.emp_sensitivity >= 5) [toxloss]`. That there's multiple integer values that represent "You take toxloss from emp" is a consequence of trying to store the different states in consecutive integers instead of as bitflags, where you'd instead get something on the order of `if(species.emp_sensitivity & EMP_GIVES_TOXLOSS) [toxloss]`. I feel I should again emphasize the single &, which is performing a logical AND of each individual bit (`0b10 & 0b11 = 0b10`) rather than performing the boolean and of "X is non-zero AND Y is non-zero". Especially when these internal values are stored as magic numbers, which you only explain in a comment next to the declaration of the base variable, rather than defines, this logic is pretty hard for someone else to understand what exactly the numbers mean, especially when you have to use the complex behaviour of "these multiple sets of possible values each trigger this one thing", as opposed to "if this bit is set, do this this"

Reply:
Yeah, it was a bit of a patch-fix when I wanted to make it a bit more flexible instead of a linear progression. The whole thing is honestly kind of messy, but I'm not sure how better to handle it unless I invert the order in a tidier fashion, such as checking weaknesses *then* applying effects multiplied by some value based off the emp severity. 🤔

Guess I'll start digging and see what I can whip up!

Reply:
..hm, you know, that might actually work. EMP intensity working in *reverse order* means that I think I can subtract the severity from a base duration/damage and then round up to zero if the result would be negative.

One thing at a time though.

Reply:
Refactored the entire thing to integrate severity in a cleaner fashion, and introduced a little more randomness to tiers. Also added support for pain (how did I overlook that?!), seperated out the stun types, and the other two damage types.

Basically the goal I'm going for here is to make this as flexible as possible, to support whatever Fun Stuff folks want to do with it.